feat: auto sync Implementation V1

Stijnus opened the pull request:
# Sync Implementation

## Overview

The sync functionality provides a robust system for synchronizing project files with a local directory. It features intelligent project folder management with timestamp-based naming, automated sync capabilities, comprehensive sync statistics tracking, and a modern UI with trend analysis.

## Key Components

### 1. Project Folder Management
- Singleton `ProjectFolderManager` for centralized folder handling
- Project name normalization (lowercase, alphanumeric with underscores)
- Timestamp-based unique folder naming (e.g., `project_name_a1b2c3`)
- Persistent storage of project mappings in localStorage
- Smart matching of existing projects (exact match and base name match)

### 2. Sync Settings
- Auto-sync with 30-second check intervals
- Sync-on-save option
- Three sync modes:
  - Ask: Prompt for each file conflict
  - Overwrite: Replace existing files without prompting
  - Skip: Preserve existing files
- Default exclude patterns: ['node_modules/**', '*.log', '.DS_Store']

### 3. Sync Process
- Real-time progress tracking with toast notifications
- Binary file handling (excluded from sync)
- Directory structure preservation
- Conflict detection and resolution based on sync mode
- Pattern-based file exclusion using 'ignore' package

### 4. Statistics & History
- Per-sync statistics tracking:
  - Number of files synced
  - Total data size
  - Sync duration
  - Timestamp
- Trend analysis comparing current vs previous periods
- Comprehensive sync history:
  - Unique UUID per sync
  - Project name
  - Status (success/partial/failed)
  - List of synced files
  - Limited to last 100 entries
  - Paginated view (10 items per page)

## UI Components

### 1. Core Components
- `SyncTab`: Main sync control interface
  - Sync folder selection and display
  - Manual sync trigger
  - Real-time sync status indicator
  - Settings configuration
  - File statistics display

- `SyncStats`: Comprehensive statistics and history view
  - Time-range filtered statistics
  - Trend indicators for key metrics
  - Visual timeline for sync history
  - Paginated history list
  - Latest sync summary

### 2. Reusable Components
- `StatsCard`: Statistics display with trend indicators
  ```typescript
  interface StatsCardProps {
    icon: string;
    label: string;
    value: string | number;
    color?: string;
    trend?: { value: number; isPositive: boolean } | null;
  }
  ```

- `SyncStatusIndicator`: Real-time sync status visualization
  ```typescript
  interface SyncStatusIndicatorProps {
    status: 'idle' | 'syncing' | 'error';
  }
  ```

- `TimeRangeSelector`: Time range filtering control
  ```typescript
  interface TimeRangeSelectorProps {
    value: string;
    onChange: (value: string) => void;
    options: Array<{ value: string; label: string }>;
  }
  ```

- `HistoryEntry`: Timeline-based history entry display
  ```typescript
  interface HistoryEntryProps {
    entry: SyncHistoryEntry;
    expanded: boolean;
    onToggle: () => void;
    formatters: {
      size: (bytes: number) => string;
      time: (date: number | Date) => string;
    };
  }
  ```

## Technical Implementation

### Core Data Structures

```typescript
interface SyncSettings {
  autoSync: boolean;
  autoSyncInterval: number;
  syncOnSave: boolean;
  excludePatterns: string[];
  syncMode: 'ask' | 'overwrite' | 'skip';
  projectFolders: Record<string, ProjectSyncInfo>;
}

interface ProjectSyncInfo {
  projectName: string;
  folderName: string;
  lastSync: number;
}

interface SyncSession {
  id: string;
  timestamp: number;
  lastSync: number;
  files: Set<string>;
  history: SyncHistoryEntry[];
  statistics: SyncStatistics[];
  projectFolder?: string;
  projectName?: string;
}
```

## Sync Process Flow

1. **Project Setup**
   - Normalize project name (lowercase, remove special chars)
   - Find existing project or create new one
   - Generate unique folder name with timestamp suffix
   - Create project directory if needed

2. **Sync Initialization**
   - Load sync settings from localStorage
   - Initialize ProjectFolderManager singleton
   - Set up 30-second auto-sync checker
   - Create or resume sync session

3. **File Synchronization**
   - Display progress toast notification
   - For each non-binary file:
     - Check against exclude patterns
     - Create necessary directories
     - Handle conflicts per sync mode
     - Write file content if approved
     - Track sync statistics

4. **History & Statistics**
   - Generate unique sync ID (UUID)
   - Record sync statistics (files, size, duration)
   - Update session history
   - Maintain last 100 entries in localStorage
   - Update last sync timestamp
   - Calculate trends for statistics

## Usage Guide

1. **Initial Setup**
   - Select root sync folder
   - Configure sync mode preference
   - Set exclude patterns if needed
   - Enable auto-sync if desired

2. **Manual Sync**
   - View current sync status via the status indicator
   - Click "Sync Now" to trigger manual synchronization
   - Monitor real-time progress with visual feedback
   - Review updated statistics and history entries
   - Handle any conflict prompts based on sync mode

3. **Auto-sync**
   - Enable "Auto Sync" in settings
   - Configure sync interval (1-3600 seconds)
   - System automatically syncs files when saved
   - Real-time status updates every 10 seconds
   - History and statistics refresh automatically

## Best Practices

1. **Project Organization**
   - Use descriptive project names
   - Let system handle folder naming
   - Review exclude patterns for efficiency

2. **Conflict Management**
   - Choose appropriate sync mode
   - Monitor partial sync status
   - Review conflict prompts carefully

3. **Performance**
   - Configure reasonable auto-sync intervals
   - Exclude unnecessary files/directories
   - Monitor sync statistics and trends
   - Clean old sync history periodically
<img width="900" alt="image" src="https://github.com/user-attachments/assets/b7b71631-ac40-4984-bf4f-3b3c4e85d814" />

<img width="891" alt="image" src="https://github.com/user-attachments/assets/a0895e8e-3b80-45d1-9882-5cd4376ce98f" />

<img width="828" alt="image" src="https://github.com/user-attachments/assets/deb560e9-5ba3-42b7-a233-a3359c0f6cc1" />

<img width="905" alt="image" src="https://github.com/user-attachments/assets/789eef52-bdba-44d2-9a26-eca5f11af3ab" />

thecodacus commented:
this is awesome!!!

thecodacus commented on SYNC_IMPLEMENTATION.md:
```diff
@@ -0,0 +1,206 @@
+# Sync Implementation
```
do we need this file or is it for drafting the PR description?

Stijnus replied:
its a draft for the PR.

thecodacus commented on app/components/ui/Dropdown.tsx:
```diff
@@ -0,0 +1,48 @@
+import { type ReactNode } from 'react';
+import { Transition } from '@headlessui/react';
+
```
since all the other components are using radix ui, can you adapt the same. otherwise it will soon bloat with multiple ui frameworks

thecodacus replied:
and is its only for animation transition then try framer motion

thecodacus commented on app/components/ui/Tooltip.tsx:
```diff
@@ -1,79 +1,29 @@
+import { type ReactNode } from 'react';
+import * as TooltipPrimitive from '@radix-ui/react-tooltip';
 
-interface TooltipProps {
```
I appreciate this update but is this change necessary, or required for your UI, if not I recommend not do this change and if its required then keep name of the component same as that might confuse others who are already working on other prs. this can cause lots of conflicts also

thecodacus commented:
looks god to me, my only recommendation is revert all the db refactoring and, tooltip changes to as it is. anc if required can be done in separate PR

leex279 commented:
I testet it out and its a pretty nice feature, but seems buggy to me.

If I select an folder via the Settings and then create one project => need to select the folder again in the project => why?

I selected it, then it synced, what is ok, but then the folder in settings is removed and cant sync anymore:
![image](https://github.com/user-attachments/assets/508ea025-7b32-4cda-b7be-48380e28b664)

Question => where to set which projects/chats should be synced? I would expect a button to activate the autosync per project/chat and/or within the settings.

Stijnus commented:
> I testet it out and its a pretty nice feature, but seems buggy to me.
> 
> If I select an folder via the Settings and then create one project => need to select the folder again in the project => why?
> 
> I selected it, then it synced, what is ok, but then the folder in settings is removed and cant sync anymore: ![image](https://private-user-images.githubusercontent.com/6664492/402375324-508ea025-7b32-4cda-b7be-48380e28b664.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3MDcyNDksIm5iZiI6MTczNjcwNjk0OSwicGF0aCI6Ii82NjY0NDkyLzQwMjM3NTMyNC01MDhlYTAyNS03YjMyLTRjZGEtYjdiZS00ODM4MGUyOGI2NjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTJUMTgzNTQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWEzMTU1MWIyYjYwZWQ5YzY3YmM5ZDRmNTg3NTJmMDI5ZjE5YjM4NDlhYjdhNGZmOGQ0NDI1YTNiOGE5NzZjMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Eqne173P39B9-_VtS8vVOnML0wJnapKeWTwlNZcq8AY)
> 
> Question => where to set which projects/chats should be synced? I would expect a button to activate the autosync per project/chat and/or within the settings.

Hi,

I have tested again the auto sync function when i set the sync folder and start a chat/project i is still enabled . Also in the workbench you can see the status of the sync 
<img width="808" alt="image" src="https://github.com/user-attachments/assets/9f99fd5b-6e87-46ed-9bf0-9ce2a1afa5fe" />

when you start a chat/project and you dont enable it in the settings tab you have the option in the workbench to enable it :
<img width="831" alt="image" src="https://github.com/user-attachments/assets/a829b3a7-3b59-47a2-885e-cc1e2b3b66d5" />
<img width="1208" alt="image" src="https://github.com/user-attachments/assets/e4798d78-1c98-42eb-bf5b-9a35d5d3ea16" />
<img width="817" alt="image" src="https://github.com/user-attachments/assets/03571fb1-4783-4143-9601-af6903deff45" />

leex279 commented:
Hi @Stijnus,

I testet again with a clean incognito browser, but with same result. I am on Windows 11. Dont think it´s an OS problem, but who knows.

**See Screencast:**
https://github.com/user-attachments/assets/15b41904-57a1-46cc-b092-4a845913e9ed


Other then that => I dont see an option to manage which projects I want to sync and what not. Should be a checkbox to enable/disable auto-sync for each project/chat.

Also to consider how to load projects which are already synced. Then I have to reimport at the moment, get a new chat and then would get another folder on the host and got it twice?

Digitl-Alchemyst commented:
This feature is a great addition and incredibly powerful for users in production environments. However, I have some concerns about the default 30-second sync interval.

Issues I see with the Sync Rate and lack of control:
- Performance Impact: On slower or older systems, frequent syncing—especially with larger repositories—could significantly degrade performance and potentially result in continuous syncing.
- Risk of Syncing Unstable Code: There’s a risk of unintentionally syncing incomplete or unstable code from Bolt to a local repository that the user might be trying to safeguard.

Suggestions for Improvement:
1. Manual Sync Option: Add a sync button in the editor, allowing users to manually trigger syncing when they’re ready.
2. Configurable Auto-Sync: Include a setting to toggle auto-sync on/off and customize the sync interval (in seconds).
3. Sync Safety Check: Ensure that Bolt AI isn’t writing files during sync operations to prevent conflicts.
4. Sync Direction Control (Bonus): Providing users with control over sync direction would be a valuable enhancement.

Let me know your thoughts on these suggestions!

thecodacus commented:
so here a re some of my thoughts or suggestion of the concerns others have

>Other then that => I dont see an option to manage which projects I want to sync and what not. Should be a checkbox to enable/disable auto-sync for each project/chat.

i dont think we should have project based on off toggle, that can make a huge list. rather an option to turn on/off auto sync on the workbench header panel for quick access.

>Also to consider how to load projects which are already synced. Then I have to reimport at the moment, get a new chat and then would get another folder on the host and got it twice?

this is a limitation even now when you sync. so I don't think this needs to be fixed in this same PR, but here are some suggestion of the fix

when syncing add a file in the `.bolt` called `config.json` if its not already present and add the chat_id, and folder name, of the project
and when we are reimporting the folder the project it can lookup the db and see if the id exist, then it can show a popup that there is already a chat with this project would you like to continue/ or start new chat session.

but in my opinion it should be a separate PR

>Risk of Syncing Unstable Code: There’s a risk of unintentionally syncing incomplete or unstable code from Bolt to a local repository that the user might be trying to safeguard.

I agree but since we have undo option available, we should be able to revert to any previous chat and the sync will automatically revert in local session also. 

> Sync Safety Check: Ensure that Bolt AI isn’t writing files during sync operations to prevent conflicts.

so this makes me  think does sync also happens when project is loading or being written, we sould pause sync when there is a reload of project or the llm is currently writing 

> Sync Direction Control (Bonus): Providing users with control over sync direction would be a valuable enhancement.

I think current sync direction is only from bolt-> local its not bidirectional, as this is only happening from browser and it has limited local storage access and restrictions. 
so i believe this PR is also not doing 2way sync

Stijnus commented:
Hello,

Thank you all for the feedback and i have applied some changes and ui fixes  : 

The tooltip should now show:

Sync folder name
Last sync time
Number of files and total size
Auto-sync status with interval (if enabled)
Sync-on-save status (if enabled)

Removed the arbitrary 30-second check interval
Added a new syncIntervalId field to track the active interval
Updated the sync settings subscription to:
Clear any existing interval when settings change
Set up a new interval that matches the user's configured interval (in minutes)
The interval now directly triggers a sync instead of checking if enough time has passed
Added proper cleanup of intervals when settings change

I will check other things tomorrow and make a new PR time for bed now :)

Digitl-Alchemyst commented:
This latest commit introduces a conflict with main that will need be resolved.
![image](https://github.com/user-attachments/assets/5069fc6a-b10d-453b-8198-7c61d029774f)

in regards to  @thecodacus reply on the bidirectional sync Yeah I wasn't sure on the sync direction stuff, I had phrased it more as a question in my original draft.  I do have a follow-up thought on that, We already have the upload entire folder feature, this feature could (now or in the future) make use of that, I am seeing such a way where the sync feature can call the upload folder and have a preset location it pulls from, although the more I think about it though the only use I see for bolt pulling form the folder, would be like on project start, maybe if you have the sync feature turned on you will get a pop up when you load the project asking if you want to pull updates from local before you start?

leex279 commented:
@Stijnus thanks, seems to be more stable now and not buggy.

But I am still missing the function disable/activate the sync for a specific chat/project, cause at least I dont want to sync always every chat. I do a lot of test in differnt chats which are just temporary but maybe I got one where I got a productive project, which I want to sync.
This is not possible at the moment or I dont see it. 

Syncing all brings a lot of waste to my sync folder I have to clean up manually then.

Stijnus commented:
> @Stijnus thanks, seems to be more stable now and not buggy.
> 
> But I am still missing the function disable/activate the sync for a specific chat/project, cause at least I dont want to sync always every chat. I do a lot of test in differnt chats which are just temporary but maybe I got one where I got a productive project, which I want to sync. This is not possible at the moment or I dont see it.
> 
> Syncing all brings a lot of waste to my sync folder I have to clean up manually then.

Hi, @leex279 let me check it tomorrow first thing ...

Digitl-Alchemyst commented:
Currently 1 test failing:
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

 Test Files  1 failed | 2 passed (3)
      Tests  25 passed (25)
   Start at  22:24:58
   Duration  1.46s (transform 511ms, setup 0ms, collect 754ms, tests 32ms, environment 1ms, prepare 270ms)


Error: ReferenceError: localStorage is not defined
 ❯ app/lib/stores/workbench.ts:111:7
 ❯ Object.subscribe node_modules/.pnpm/nanostores@0.10.3/node_modules/nanostores/atom/index.js:77:7
 ❯ new WorkbenchStore app/lib/stores/workbench.ts:110:23
 ❯ app/lib/stores/workbench.ts:[98](https://github.com/stackblitz-labs/bolt.diy/actions/runs/12757224053/job/35557031681?pr=1076#step:5:99)8:31
 ❯ app/components/chat/Artifact.tsx:10:31


close timed out after [100](https://github.com/stackblitz-labs/bolt.diy/actions/runs/12757224053/job/35557031681?pr=1076#step:5:101)00ms
Tests closed successfully but something prevents Vite server from exiting
You can try to identify the cause by enabling "hanging-process" reporter. See https://vitest.dev/config/#reporters
 ELIFECYCLE  Test failed. See above for more details.
Error: Process completed with exit code 1.

Stijnus commented:
@leex279 i have made some changes in to https://github.com/stackblitz-labs/bolt.diy/pull/1092 and fix other issues. I hope you like them...   😎
I will post a vid soon in the V2 PR.

Br,
Stijn(us)

leex279 commented:
will checkout tomorrow. thanks :)

Digitl-Alchemyst commented:
@Stijnus is the new PR replacing this one?

Stijnus commented:
> @Stijnus is the new PR replacing this one?

Yes Its V2 of the "Auto Sync"  : https://github.com/stackblitz-labs/bolt.diy/pull/1092

Br,

Stijn

leex279 commented:
=> Closed, cause #1092 replaces this one
